### ISSUE TITLE
unresolved import: `thiserror::Error`

### ISSUE THREAD

**Issue by bryanburgers** [rust-analyzer/rust-analyzer#6053]
In the latest version of rust-analyzer (on nvim+coc, but a coworker is reporting the same thing on VSCode), I'm getting the following error from rust-analyzer.

<img width="409" alt="image" src="https://user-images.githubusercontent.com/56045/93794545-d3630100-fbfd-11ea-9b6a-f195a76bcf5d.png">

`[rust-analyzer] [E] unresolved import`

This happens even though a regular build of the project finishes just fine.

[`thiserror::Error`] is a derive macro without an accompanying trait (c.f. `serde::Serialize`, which is a derive macro _and also_ a trait, which does not cause an error).

Could there be a regression with importing derive macro traits?

[`thiserror::Error`]: https://docs.rs/thiserror/1.0.20/thiserror/derive.Error.html

**Comment by pluth** [comment]
I'm having a similar issue with snafu::Snafu (another error-handling derive macro)


![Screenshot from 2020-09-21 18-41-15](https://user-images.githubusercontent.com/39027073/93795750-1d69d800-fc3a-11ea-8168-d9be7af0c41f.png)

**Comment by bjorn3** [comment]
You can enable proc-macro support using

```jsonc
{
    "rust-analyzer.cargo.loadOutDirsFromCheck": true,
    "rust-analyzer.procMacro.enable": true,
}
```

**Comment by Systemcluster** [comment]
I had the same issue with derive macro imports after the latest update with an existing project, but cleaning the target, rebuilding the project and reloading the editor solved it.

**Comment by pluth** [comment]
> You can enable proc-macro support using
> 
> ```js
> {
>     "rust-analyzer.cargo.loadOutDirsFromCheck": true,
>     "rust-analyzer.procMacro.enable": true,
> }
> ```

I thought, I had this enabled but apparently it was a workspace setting :facepalm: . Fixed for me at least

**Comment by EverlastingBugstopper** [comment]
> You can enable proc-macro support using
>
> ```
> {
>     "rust-analyzer.cargo.loadOutDirsFromCheck": true,
>     "rust-analyzer.procMacro.enable": true,
> }
>```

This worked for me! Is there any reason this isn't enabled by default? It seems like this is a fairly common use case/language feature.

**Comment by lnicola** [comment]
It can be a security risk (you might load a project with a malicious build scripts), it slows down the first load of the project (because we need to compile them), it doesn't always work (because of ABI differences, e.g. proc macros don't work in nightly and beta right now and we can't fix them without breaking stable users).

**Comment by bjorn3** [comment]
> It can be a security risk (you might load a project with a malicious build scripts)

That already happens with `cargo check`.

**Comment by kellpossible** [comment]
> You can enable proc-macro support using

Enabling those options worked for me but they made RA super slow.

**Comment by bjorn3** [comment]
You can use `"rust-analyzer.diagnostics.disabled": ["unresolved-import"]` to disable the errors for this. It is already disabled in the latest version by default I think. (https://github.com/rust-analyzer/rust-analyzer/pull/6085)

**Comment by sunjay** [comment]
I seem to be getting the unresolved import error for `thiserror::Error` regardless of whether I have `procMacro` enabled. I am on `rust-analyzer version: nightly (7b674f9)`.

**Comment by bjorn3** [comment]
Do you also have `"rust-analyzer.cargo.loadOutDirsFromCheck": true`?

**Comment by sunjay** [comment]
Yes. Sorry, I should have posted my config:

```json
{
    "rust-analyzer.updates.channel": "nightly",
    "rust-analyzer.procMacro.enable": true,
    "rust-analyzer.cargo.loadOutDirsFromCheck": true,
    "rust-analyzer.inlayHints.parameterHints": false,
}
```

This just started happening today when I updated to the latest nightly. (My previous comment has the version)

**Comment by Ten0** [comment]
Same here. Additionally, it seems that cargo will stop checking at all if ` "rust-analyzer.cargo.loadOutDirsFromCheck": true,`. (Though that may be more related to the fact my project is large?)
Note I didn't have these options enabled before, but enabling them did not fix this issue for me, it still shows the unresolved imports.

**Comment by lnicola** [comment]
@sunjay 

> I seem to be getting the unresolved import error for thiserror::Error regardless of whether I have procMacro enabled. I am on rust-analyzer version: nightly (7b674f9).

https://github.com/rust-analyzer/rust-analyzer/pull/5651

**Comment by Count-Count** [comment]
I worked around this issue thus:
`
use thiserror::*; // should be thiserror::Error, but https://github.com/rust-analyzer/rust-analyzer/issues/6053
`

**Comment by GopherJ** [comment]
![image](https://user-images.githubusercontent.com/33961674/94996852-504f8c80-05a7-11eb-8b3e-61b8654b149a.png)
I'm having similar issue

**Comment by matklad** [comment]
cc @jonas-schievink

**Comment by jonas-schievink** [comment]
They're all working now :tada: (the `procMacro.enable` setting can be kept off, we can now resolve proc macro in rust-analyzer itself without consulting Cargo)

**Comment by teohhanhui** [comment]
Now we only need a stable release.

**Comment by flodiebold** [comment]
The fix for this has been included in the last two stable releases, as far as I can see.

**Comment by teohhanhui** [comment]
> rust-analyzer version: 2020-10-19 (3ca97b0)

This is still happening to me with things like:

```rust
use derivative::Derivative;
use strum_macros::EnumIter;
```

Is that a separate issue or should this be reopened?

**Comment by lnicola** [comment]
@teohhanhui did you enable build script and proc macro support as in https://github.com/rust-analyzer/rust-analyzer/issues/6053#issuecomment-699028778?

**Comment by flodiebold** [comment]
`strum_macros` was mentioned [here](https://github.com/rust-analyzer/rust-analyzer/pull/6143#issuecomment-714119369) as well and seems to be related to `cfg_attr` support, which we already have #5548 for. It'd be worth verifying that for both of these though.

**Comment by teohhanhui** [comment]
@lnicola:

> did you enable build script and proc macro support as in [#6053 (comment)](https://github.com/rust-analyzer/rust-analyzer/issues/6053#issuecomment-699028778)?

https://github.com/rust-analyzer/rust-analyzer/issues/6053#issuecomment-708590486

> the `procMacro.enable` setting can be kept off, we can now resolve proc macro in rust-analyzer itself without consulting Cargo

`rust-analyzer.cargo.loadOutDirsFromCheck` is already set to `true`.

**Comment by teohhanhui** [comment]
I think this has resolved the issue for me:

```
cargo clean
cargo build
```

**Comment by itmecho** [comment]
I'm also getting this error in neovim with the built in lsp client but it's happening on a trait from the std lib rather than an external library. The autocompeltion also seems to not be working for the `std::os` module.

![screenshot](https://i.imgur.com/lWNTWoh.png)

My nvim_lsp config:
```
local nvim_lsp = require('nvim_lsp')
local on_attach = function(client)
    require'completion'.on_attach(client)
    require'diagnostic'.on_attach(client)
end

...

nvim_lsp.rust_analyzer.setup{
    on_attach = on_attach;
    settings = {
        ["rust-analyzer"] = {
            cargo = {
                loadOutDirsFromCheck = true;
            }
        }
    }
}
```

**Comment by bjorn3** [comment]
That is a duplicate of https://github.com/rust-analyzer/rust-analyzer/issues/6038.

**Comment by itmecho** [comment]
Ah thanks, I'll follow that issue as well

**Comment by wpbrown** [comment]
> `strum_macros` was mentioned [here](https://github.com/rust-analyzer/rust-analyzer/pull/6143#issuecomment-714119369) as well and seems to be related to `cfg_attr` support, which we already have #5548 for. It'd be worth verifying that for both of these though.

Indeed derivative is the same issue. Thanks

```rust
#[cfg_attr(not(test), proc_macro_derive(Derivative, attributes(derivative)))]
pub fn derivative(input: TokenStream) -> TokenStream {
```

**Comment by hammypants** [comment]
This is still happening for the original issue on latest nightly.

**Comment by bjorn3** [comment]
Which rustc version are you using?

**Comment by hammypants** [comment]
> Which rustc version are you using?

`rustc 1.46.0 (04488afe3 2020-08-24)`

I'll try updating to latest.

**Comment by aaomidi** [comment]
![image](https://user-images.githubusercontent.com/1465785/105921725-3c70e000-6007-11eb-9b27-6f4afbf8e51f.png)
Not working for me unfortunately.


![image](https://user-images.githubusercontent.com/1465785/105921743-4266c100-6007-11eb-9c66-0eb7a3bd37a5.png)

Settings are on true

**Comment by bjorn3** [comment]
I think there was an abi breaking change on rustc nightly. Either switch to stable or wait for this change to land on stable such that rust-analyzer can adapt to it.

**Comment by edwin0cheng** [comment]
@aaomidi Confirm your bug in `rustc-stable`, 

maybe related to `cfg_attr(not(test)...)`  ?

https://github.com/mcarton/rust-derivative/blob/770c67484a872470ca782202e13fc70a35e2695b/src/lib.rs#L76-L80

**Comment by Folaht** [comment]
rust-ls.lua

```
require'lspconfig'.rust_analyzer.setup{
    cmd = {DATA_PATH .. "/lspinstall/rust/rust-analyzer"},
    on_attach = require'lsp'.common_on_attach,
	settings = {
		["rust-analyzer"] = {
			rustprocMacro = {
				enable = true
			},
			cargo = {
				loadOutDirsFromCheck = true
			},
		}
	}
}
```

**Comment by lnicola** [comment]
@Folaht your `rustprocMacro` key is misspelled (see the [manual](https://rust-analyzer.github.io/manual.html#nvim-lsp)), but these should be enabled by default in recent versions.

**Comment by OfekShochat** [comment]
Im getting this currently

**Comment by lnicola** [comment]
@OfekShochat that's probably https://github.com/rust-analyzer/rust-analyzer/issues/7243

**Comment by OfekShochat** [comment]
yes it is, looked at it

**Comment by aranw** [comment]
I tried to add the suggested fixes above in this issue but I'm still getting the same issue. Is there something else that I am missing? 

```
❯ rustc --version
rustc 1.70.0-nightly (39f2657d1 2023-03-09)
```

I'm also using  rust-analyzer `v0.4.1582`
